pkg-manager/resolve-dependencies/src/resolvePeers.ts

```diff
@@ -110,13 +114,67 @@ export function resolvePeers<T extends PartialResolvedPackage> (
   for (const { directNodeIdsByAlias, id } of opts.projects) {
     dependenciesByProjectId[id] = mapValues((nodeId) => pathsByNodeId[nodeId], directNodeIdsByAlias)
   }
+  if (opts.dedupePeerDependents) {
```

Comment:
It looks like this "patches" the resolution after it's occurred.

That works, but it may be worth discussing an alternative approach that behaves similarly to `resolve-peers-from-workspace-root`, but instead passes in the entire `depGraph` rather than only workspace root dependencies. Would that work as well?

I'm wondering if there's a way we an avoid recomputing the dep graph again, and have it deduped already when `peersOfChildren` is called.

Reply:
I assume you mean to move the dedupe logic to this file: https://github.com/pnpm/pnpm/blob/main/pkg-manager/resolve-dependencies/src/resolveDependencies.ts

I was thinking about it. It was my first idea. However, at that point we don't know yet that the packages are duplicated (we know maybe the ones that were duplicated during previous install). The packages with peer dependencies are duplicated only after peers resolution, which is the next stage. So we can only deduplicate them after peers resolution.

My first idea was to try to tweak the code that does auto installation of peer dependencies and prefer versions that are already present in the lockfile or are already used to resolve peers. But that was a lot harder to implement and less reliable I believe.

Reply:
> I assume you mean to move the dedupe logic to this file: https://github.com/pnpm/pnpm/blob/main/pkg-manager/resolve-dependencies/src/resolveDependencies.ts

Roughly. To elaborate more, I was thinking of adding the full dep graph to the `pkgsByName` object above, in addition to  `rootPkgsByName`.

> I was thinking about it. It was my first idea. However, at that point we don't know yet that the packages are duplicated (we know maybe the ones that were duplicated during previous install). The packages with peer dependencies are duplicated only after peers resolution, which is the next stage. So we can only deduplicate them after peers resolution.

That's true. Adding more available peers to select from during peers resolution would change the result without knowing whether that change was necessary to dedupe. I find something nice about being able to intuit "_If it's in the lockfile, it can be a peer._", but I could also see that being undesirable for complicated cases.

Don't have strong opinions here. Your final solution is good. 👍

Thanks for discussing!